service/common/src/main/java/org/apache/polaris/service/catalog/iceberg/IcebergCatalog.java

```diff
@@ -881,11 +859,57 @@ public Map<String, String> getCredentialConfig(
         getCredentialVendor(),
         callContext.getPolarisCallContext().getConfigurationStore(),
         tableIdentifier,
-        getLocationsAllowedToBeAccessed(tableMetadata),
+        locationsAllowedToBeAccessed,
         storageActions,
         storageInfo.get());
   }
 
+  public MetadataJson loadTableMetadataJson(TableIdentifier identifier) {
+    return loadTableMetadataJson(identifier, SNAPSHOTS_ALL);
+  }
+
+  boolean shouldFilterSnapshots(String snapshots) {
+    if (snapshots == null || snapshots.equalsIgnoreCase(SNAPSHOTS_ALL)) {
+      return false;
+    }
+    return callContext
+        .getPolarisCallContext()
+        .getConfigurationStore()
+        .getConfiguration(
+            callContext.getPolarisCallContext(), FeatureConfiguration.ALWAYS_FILTER_SNAPSHOTS);
+  }
+
+  public MetadataJson loadTableMetadataJson(TableIdentifier identifier, String snapshots) {
+    int maxMetadataCacheBytes =
+        callContext
+            .getPolarisCallContext()
+            .getConfigurationStore()
+            .getConfiguration(
+                callContext.getPolarisCallContext(), FeatureConfiguration.METADATA_CACHE_MAX_BYTES);
+    Supplier<TableMetadata> fallback =
```

Comment:
Do we need to anonymously declare the fallback here? I'm also wondering a bit from the usage here whether we need a Supplier or Function<TableIdentifier, TableMetadata> ?

With that definition we wouldn't be generating a new anonymous function for each identifier and could have more static definitions like

TableMetadata loadFromCatalog(TableIdentifier)
TableMetadata alwaysFailFallback(TableIdentifier) <- Like in the tests

Reply:
I used `Supplier` to make testing easier as you noted, but also to capture the fact that we really may not always be loading from a catalog. For example, in the future if we directly stream metadata.json content from object storage to the client, we'd need a different kind of fallback implementation. Is the concern about the performance overhead of the lambda?

Reply:
I have no problem with the functional interface, I was just wondering if it should be a single arg function rather than a no-arg supplier. If we have a one arg function we can avoid some garbage generation here by having to make so many different suppliers. We would still support a "no-arg" supplier because it could just ignore the arg passed through.